### PR TITLE
adding Expect: header

### PR DESCRIPTION
**Pull request by vhvb1989**
fixes: https://github.com/Azure/azure-sdk-for-c/issues/615

Making curl transport adapter to automatically sets header "Expect:"
This makes libcurl to disable its behavior of sending headers only to Server and expect a Continue response to start uploading for PUT requests or POST with payload greater than 1024.

### CODE REVIEW

File: sdk/platform/http_client/curl/src/az_curl.c [hunk 1]
```diff
@@ -519,18 +546,20 @@ static AZ_NODISCARD az_result _az_http_client_curl_send_request_impl_process(
...
-  {
-    result = _az_http_client_curl_send_post_request(p_curl, p_request);
-  }
   else if (az_span_is_content_equal(p_request->_internal.method, az_http_method_delete()))
   {
     result = _az_http_client_curl_send_delete_request(p_curl, p_request);
   }
+  else if (az_span_is_content_equal(p_request->_internal.method, az_http_method_post()))
+  {
+    AZ_RETURN_IF_FAILED(_az_http_client_curl_add_expect_header(p_curl, &p_list));
```
Comment (gilbertw) [inline sdk/platform/http_client/curl/src/az_curl.c:555]: Why is this not on every request?  Curl is unclear on where they will add "Expect: 100-continue" as a header

Comment (vhvb1989) [reply sdk/platform/http_client/curl/src/az_curl.c:555]: it will be only always for PUT or for any POST with payload greater than 1024

File: sdk/platform/http_client/curl/src/az_curl.c [hunk 2]
```diff
@@ -150,7 +150,7 @@ static AZ_NODISCARD az_result _az_http_client_curl_add_header_to_curl_list(
   az_span writable_buffer;
   {
     int32_t const buffer_size = az_span_size(header.key) + az_span_size(separator)
-        + az_span_size(header.value) + az_span_size(AZ_SPAN_FROM_STR("\0"));
+        + az_span_size(header.value) + 1 /*one for 0 terminated*/;
```
Comment (ahsonkhan) [inline sdk/platform/http_client/curl/src/az_curl.c:153]: nit: In other places, folks have preferred using `az_span_size` to be more self documenting rather than constants. We may want to converge and use the same for consistency.

https://github.com/Azure/azure-sdk-for-c/pull/568#discussion_r414133370

So, for now, I would leave things as is and revert this change. It is unrelated to the fix anyway.

File: sdk/platform/http_client/curl/src/az_curl.c [hunk 3]
```diff
@@ -170,6 +170,45 @@ static AZ_NODISCARD az_result _az_http_client_curl_add_header_to_curl_list(
   return result;
 }
 
+/**
+ * @brief Adds special header "Expect:" for libcurl to avoid sending only headers to server and wait
+ * for a 100 Continue response before sending a PUT method
```
Comment (ahsonkhan) [inline sdk/platform/http_client/curl/src/az_curl.c:175]:
```suggestion
 * for a 100 Continue response before sending a PUT method.
```

File: sdk/platform/http_client/curl/src/az_curl.c [hunk 4]
```diff
@@ -170,6 +170,45 @@ static AZ_NODISCARD az_result _az_http_client_curl_add_header_to_curl_list(
...
+ *
+ * However, many servers don't implement the Expect: stuff properly and if the
+ * server doesn't respond (positively) within 1 second libcurl will continue
+ * and send off the data anyway.
+ *
+ * You can disable libcurl's use of the Expect: header the same way you disable
+ * any header, using -H / CURLOPT_HTTPHEADER, or by forcing it to use HTTP 1.0.
+ *
+ * This function is meant to be called after all headers from original request was called. It will
+ * append another header and set headers for a p_curl session
```
Comment (ahsonkhan) [inline sdk/platform/http_client/curl/src/az_curl.c:192]:
```suggestion
 * append another header and set headers for a p_curl session.
```

File: sdk/platform/http_client/curl/src/az_curl.c [hunk 5]
```diff
@@ -170,6 +170,45 @@ static AZ_NODISCARD az_result _az_http_client_curl_add_header_to_curl_list(
...
+ *
+ * @param p_curl reference to an easy curl session
+ * @param p_list list of headers as curl list
```
Comment (ahsonkhan) [inline sdk/platform/http_client/curl/src/az_curl.c:195]: nit: Reminder to use full sentences in docs (capitalized with period).

This is an internal/private API anyway, so not as important.

File: sdk/platform/http_client/curl/src/az_curl.c [hunk 6]
```diff
@@ -170,6 +170,45 @@ static AZ_NODISCARD az_result _az_http_client_curl_add_header_to_curl_list(
...
+ *
+ * @return az_result
```
Comment (ahsonkhan) [inline sdk/platform/http_client/curl/src/az_curl.c:197]: Do we want to list the set of az_results it can return and under what conditions? Similar to:
https://github.com/Azure/azure-sdk-for-c/blob/9c06eac1025670beb6cd4ccf0f0ae403e5f73dad/sdk/core/core/inc/az_span.h#L380-L385

File: sdk/platform/http_client/curl/src/az_curl.c [hunk 7]
```diff
@@ -519,18 +559,20 @@ static AZ_NODISCARD az_result _az_http_client_curl_send_request_impl_process(
...
   }
-  else if (az_span_is_content_equal(p_request->_internal.method, az_http_method_post()))
-  {
-    result = _az_http_client_curl_send_post_request(p_curl, p_request);
-  }
   else if (az_span_is_content_equal(p_request->_internal.method, az_http_method_delete()))
   {
     result = _az_http_client_curl_send_delete_request(p_curl, p_request);
   }
+  else if (az_span_is_content_equal(p_request->_internal.method, az_http_method_post()))
```
Comment (ahsonkhan) [inline sdk/platform/http_client/curl/src/az_curl.c:566]: Why was this else if branch moved to  a different order? Was this intentional and does the behavior change because of it?

File: sdk/platform/http_client/curl/src/az_curl.c (hunk 2 above)
Comment (vhvb1989) [reply sdk/platform/http_client/curl/src/az_curl.c:153]: this was a @JeffreyRichter  request. He didn't like to see the az_span created just to calculate the size of it when we know it will be just 1.

File: sdk/platform/http_client/curl/src/az_curl.c (hunk 7 above)
Comment (vhvb1989) [reply sdk/platform/http_client/curl/src/az_curl.c:566]: intentional to have PUT and POST together.   Nothing really important
